### PR TITLE
Introduce NodeJS Gapic code generator.

### PR DESCRIPTION
**Pull request by jmuk**

### CODE REVIEW

**Comment by jmuk** [comment]
cc: @tbetbetbe 

This actually does not work with current gax -- I'll create a PR there. Especially, this follows the style of https://github.com/googleapis/gax-python/issues/104 suggested by @garrettjonesgoogle, but gax-nodejs does not support this style yet.

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicProvider.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.ArrayList;
+
+/**
+ * The GapicProvider which runs Gapic code generation for Node.JS.
+ */
+public class NodeJSGapicProvider<InputElementT extends ProtoElement>
+    implements GapicProvider<InputElementT> {
+
+  private final NodeJSGapicContext context;
+  private final NodeJSProvider provider;
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicProvider.java]: I think you want NodeJSSnippetSetRunner here

Comment (jmuk) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicProvider.java]: Yeah, my working revision was a bit old. Rebased and uploaded the new patch.

File: src/main/java/com/google/api/codegen/nodejs/JSDocCommentFixer.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  /**
+   * Returns a JSDoc-formatted comment string.
+   */
+  public static String jsdocify(String comment) {
+    comment = jsdocifyProtoMarkdownLinks(comment);
+    comment = jsdocifyCloudMarkdownLinks(comment);
+    return cleanupTrailingWhitespaces(comment);
+  }
+
+  private static String protoToRubyDoc(String comment) {
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/nodejs/JSDocCommentFixer.java]: Update method name

File: src/main/java/com/google/api/codegen/nodejs/JSDocCommentFixer.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    boolean isFirstSegment = true;
+    String result = "";
+    for (String name : Splitter.on(".").splitToList(comment)) {
+      char firstChar = name.charAt(0);
+      if (Character.isUpperCase(firstChar)) {
+        messageFound = true;
+        result += (isFirstSegment ? "" : "::") + name;
+      } else if (messageFound) {
+        // Lowercase segment after message is found is field.
+        // In Ruby, it is referred as "Message#field" format.
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/nodejs/JSDocCommentFixer.java]: Update reference to Ruby

File: src/main/java/com/google/api/codegen/nodejs/JSDocCommentFixer.java (hunk 2 above)
Comment (geigerj) [inline src/main/java/com/google/api/codegen/nodejs/JSDocCommentFixer.java]: Actually, I don't see where this method is being used -- if it's not, remove it?

File: src/main/java/com/google/api/codegen/nodejs/JSDocCommentFixer.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    }
+    do {
+      String url = "https://cloud.google.com" + m.group(2);
+      m.appendReplacement(sb, String.format("[%s](%s)", m.group(1), url));
+    } while (m.find());
+    m.appendTail(sb);
+    return sb.toString();
+  }
+
+  private static String cleanupTrailingWhitespaces(String comment) {
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/nodejs/JSDocCommentFixer.java]: Why does this need to be a separate method?

(I see that this probably comes from Ruby, but there too, this method seems only to wrap `trim()`)

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+  public List<String> defaultComments(ProtoElement element) {
+    if (!element.hasAttribute(ElementDocumentationAttribute.KEY)) {
+      return ImmutableList.<String>of();
+    }
+    return convertToCommentedBlock(JSDocCommentFixer.jsdocify(
+        DocumentationUtil.getScopedDescription(element)));
+  }
+
+  /**
+   * Returns type information for a field in YARD style.
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: YARD?

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+    if (!Strings.isNullOrEmpty(paramComment)) {
+      paramComment = JSDocCommentFixer.jsdocify(paramComment);
+      comment += "\n  " + paramComment.replaceAll("(\\r?\\n)", "\n  ");
+    }
+    return comment + "\n";
+  }
+
+  /**
+   * Return JSDoc return type string for the given method, or null if the return type is
+   * nil.
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: nil?

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+        .getMethodConfig(method);
+
+    String callbackType =
+        isEmpty ? "EmptyCallback" : String.format("APICallback<%s>", classInfo);
+    String callbackComment = "@param {?" + callbackType + "} callback\n"
+        + "  The function which will be called with the result of the API call.";
+    if (config.isPageStreaming()) {
+      String resourceType = jsTypeName(config.getPageStreaming().getResourcesField().getType());
+      return callbackComment + "\n@returns {?Stream<" + resourceType + ">}\n"
+          + "  An object stream of " + resourceType + " instances unless\n"
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: "instances unless" --> "instances, unless" (add comma)

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+          + "  page streaming is disabled through the call options or callback\n"
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: "or callback" --> "or a callback"

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+          + "  is specified. If page streaming is disabled or callback is specified,\n"
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: "or callback" --> "or a callback"

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java [hunk 10]
```diff
@@ -0,0 +1,427 @@
...
+          + "  this return a null and callback will be called with a single instance\n"
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: "this return a null and callback" --> "this returns null, and the callback"

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java [hunk 11]
```diff
@@ -0,0 +1,427 @@
...
+    }
+    return name;
+  }
+
+  /**
+   * Return the default value for the given field.
+   */
+  public String defaultValue(Field field) {
+    TypeRef type = field.getType();
+    // Return empty array if the type is repeated.
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: This comment appears to be in the wrong place

File: src/main/resources/com/google/api/codegen/nodejs/main.snip [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+    @join collectionConfig : context.getApiConfig.getInterfaceConfig(service).getCollectionConfigs()
+
+      var {@pathTemplateName(collectionConfig)} = new gax.PathTemplate(
+          '{@collectionConfig.getNamePattern}');
+    @end
+
+    @join collectionConfig : context.getApiConfig.getInterfaceConfig(service).getCollectionConfigs()
+
+      {@createResourceFunction(service, collectionConfig)}
+    @end
```
Comment (geigerj) [inline src/main/resources/com/google/api/codegen/nodejs/main.snip]: Can you add the matcher functions (#134)?

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java [hunk 13]
```diff
@@ -0,0 +1,427 @@
...
+        return "String";
+      default:
+        // Numeric types and enums.
+        return "Number";
+    }
+  }
+
+  /**
+   * Convert the content string into a commented block that can be directly printed out in the
+   * generated Ruby files.
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: Update Ruby ref

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java [hunk 14]
```diff
@@ -0,0 +1,427 @@
...
+    return convertToCommentedBlock(contentBuilder.toString());
+  }
+
+  /**
+   * Return the list of messages within element which should be documented in NodeJS.
+   */
+  public ImmutableList<MessageType> filterDocumentingMessages(ProtoContainerElement element) {
+    ImmutableList.Builder<MessageType> builder = ImmutableList.builder();
+    for (MessageType msg : element.getMessages()) {
+      // Doesn't have to document map entries in Ruby because Hash is used.
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: Update Ruby ref

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java [hunk 15]
```diff
@@ -0,0 +1,427 @@
...
+
+    contentBuilder.append(
+        "\n@throws an error if the RPC is aborted.");
+    return convertToCommentedBlock(contentBuilder.toString());
+  }
+
+  /**
+   * Return the list of messages within element which should be documented in NodeJS.
+   */
+  public ImmutableList<MessageType> filterDocumentingMessages(ProtoContainerElement element) {
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: Where is this used?

**Comment by geigerj** [comment]
Did a first pass.

File: src/main/java/com/google/api/codegen/nodejs/JSDocCommentFixer.java (hunk 2 above)
Comment (jmuk) [inline src/main/java/com/google/api/codegen/nodejs/JSDocCommentFixer.java]: removed -- thanks for catching

File: src/main/java/com/google/api/codegen/nodejs/JSDocCommentFixer.java (hunk 3 above)
Comment (jmuk) [inline src/main/java/com/google/api/codegen/nodejs/JSDocCommentFixer.java]: the entire function was removed

File: src/main/java/com/google/api/codegen/nodejs/JSDocCommentFixer.java (hunk 4 above)
Comment (jmuk) [inline src/main/java/com/google/api/codegen/nodejs/JSDocCommentFixer.java]: removed and just calling `trim()` -- and yes, that's based on Ruby implementation as you expected.

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java (hunk 5 above)
Comment (jmuk) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: fixed -- JSDoc (YARD is for ruby documenting format)

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java (hunk 6 above)
Comment (jmuk) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: Updated the comment to reflect the actual behavior of the method, and the mention of nil is gone.

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java (hunk 7 above)
Comment (jmuk) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: done

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java (hunk 8 above)
Comment (jmuk) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: done

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java (hunk 9 above)
Comment (jmuk) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: done

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java (hunk 10 above)
Comment (jmuk) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: done

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java (hunk 11 above)
Comment (jmuk) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: done

File: src/main/resources/com/google/api/codegen/nodejs/main.snip (hunk 12 above)
Comment (jmuk) [inline src/main/resources/com/google/api/codegen/nodejs/main.snip]: added

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java (hunk 13 above)
Comment (jmuk) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: done

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java (hunk 14 above)
Comment (jmuk) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: removed the method itself

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java (hunk 15 above)
Comment (jmuk) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: well -- no. it was copied from RubyGapicContext and forgotten to be removed.

File: src/test/java/com/google/api/codegen/testdata/nodejs_library.baseline [hunk 16]
```diff
@@ -0,0 +1,862 @@
...
+ *   The function which will be called with the result of the API call.
+ * @throws an error if the RPC is aborted.
+ */
+LibraryServiceApi.prototype.publishSeries = function publishSeries(
+    shelf,
+    books,
+    otherArgs,
+    options,
+    callback) {
+  var args = arguejs({
```
Comment (geigerj) [inline src/test/java/com/google/api/codegen/testdata/nodejs_library.baseline]: From https://docs.omniref.com/js/npm/arguejs/0.1.2#label-Getting+started:

> you will not use conventional parameters definition as you used before. Actually, all your methods should be defined without them.

Is this syntax correct?

**Comment by geigerj** [comment]
One question about codegen output. Code generator classes LGTM.

File: src/test/java/com/google/api/codegen/testdata/nodejs_library.baseline (hunk 16 above)
Comment (jmuk) [inline src/test/java/com/google/api/codegen/testdata/nodejs_library.baseline]: Removed function arguments.

I actually imitated this structure from gRPC node code, and the main motivation here was for JSDoc -- I had a concern that JSDoc might got confused by the @param mismatching with actual arguments.

I tried to run jsdoc, and that concern was wrong and JSDoc still produces documents for the arguments from the doc-comment.

**Comment by geigerj** [comment]
LGTM

**Comment by jmuk** [comment]
Updated the code a bit to work properly with current gax-nodejs -- without googleapis/gax-python#104.
PTAL

**Comment by jmuk** [comment]
Okay, I'm going to merge this.

**Pull request merged**